samples/subsys/ipc/openamp/remote/src/main.c

```diff
@@ -0,0 +1,229 @@
+/*
+ * Copyright (c) 2018, NXP
+ * Copyright (c) 2018, Nordic Semiconductor ASA
+ * Copyright (c) 2018, Linaro Limited
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr.h>
+#include <ipm.h>
+#include <misc/printk.h>
+#include <device.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include <openamp/open_amp.h>
+#include <metal/device.h>
+
+#include "common.h"
+
+#define APP_TASK_STACK_SIZE (1024)
+K_THREAD_STACK_DEFINE(thread_stack, APP_TASK_STACK_SIZE);
+static struct k_thread thread_data;
+
+static struct device *ipm_handle;
+
+static metal_phys_addr_t shm_physmap[] = { SHM_START_ADDR };
+static struct metal_device shm_device = {
+	.name = SHM_DEVICE_NAME,
+	.bus = NULL,
+	.num_regions = 1,
+	{
+		{
+			.virt       = (void *) SHM_START_ADDR,
+			.physmap    = shm_physmap,
+			.size       = SHM_SIZE,
+			.page_shift = 0xffffffff,
+			.page_mask  = 0xffffffff,
+			.mem_flags  = 0,
+			.ops        = { NULL },
+		},
+	},
+	.node = { NULL },
+	.irq_num = 0,
+	.irq_info = NULL
+};
+
+static volatile unsigned int received_data;
+
+static struct virtio_vring_info rvrings[2] = {
+	[0] = {
+		.info.align = VRING_ALIGNMENT,
+	},
+	[1] = {
+		.info.align = VRING_ALIGNMENT,
+	},
+};
+static struct virtio_device vdev;
+static struct rpmsg_virtio_device rvdev;
+static struct metal_io_region *io;
+static struct virtqueue *vq[2];
+
+static unsigned char virtio_get_status(struct virtio_device *vdev)
+{
+	return sys_read8(OPENAMP_STATUS_ADDR);
+}
+
+static u32_t virtio_get_features(struct virtio_device *vdev)
+{
+	return 1 << VIRTIO_RPMSG_F_NS;
+}
+
+static void virtio_set_features(struct virtio_device *vdev,
+				u32_t features)
+{
+}
+
+static void virtio_notify(struct virtqueue *vq)
+{
+	u32_t dummy_data = 0x00110011; /* Some data must be provided */
+
+	ipm_send(ipm_handle, 0, 0, &dummy_data, sizeof(dummy_data));
+}
+
+struct virtio_dispatch dispatch = {
+	.get_status = virtio_get_status,
+	.get_features = virtio_get_features,
+	.set_features = virtio_set_features,
```

Comment:
I believe that the `.set_features` callback is not needed for the remote:
https://github.com/OpenAMP/open-amp/blob/master/lib/remoteproc/remoteproc_virtio.c#L167

Reply:
Or maybe there is an inconsistency in the OpenAMP implementation:
https://github.com/OpenAMP/open-amp/issues/139

Reply:
For now, as RPMsg is the only virtio user, and the demo doesn't make use of .set_features on the remote side. Implementation is not required. But prefer to have comments to the virtio dispatch ops structure to explain why there is no implementation for those functions.